# Предприятия общественного питания в Москве

Учебный проект по специальности "аналитик данных", 21-я неделя обучения.
Тема: Сторителлинг по данным.

Автор: Александр Лагутин.

Начат: 25 октября 2022 г в 08:30.
Завершён: 0 месяца 2022 г в 00:00.

Время подготовки: 0 рабочих часов (0 рабочих дня).

- 2 часа - оформление плана работ.
- 2 часа 30 минут - удаление неявных повторов.
- 1 час - другая подготовка данных.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Заказчик" data-toc-modified-id="Заказчик-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Заказчик</a></span></li><li><span><a href="#Задача" data-toc-modified-id="Задача-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Задача</a></span></li><li><span><a href="#Входные-данные" data-toc-modified-id="Входные-данные-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Входные данные</a></span><ul class="toc-item"><li><span><a href="#Термины" data-toc-modified-id="Термины-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Термины</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Описание данных</a></span><ul class="toc-item"><li><span><a href="#Предприятия-питания-places" data-toc-modified-id="Предприятия-питания-places-1.3.2.1"><span class="toc-item-num">1.3.2.1&nbsp;&nbsp;</span>Предприятия питания <code>places</code></a></span></li></ul></li></ul></li><li><span><a href="#План-работы" data-toc-modified-id="План-работы-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>План работы</a></span></li><li><span><a href="#Инструменты" data-toc-modified-id="Инструменты-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Инструменты</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Определения-констант" data-toc-modified-id="Определения-констант-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Определения констант</a></span></li><li><span><a href="#Определения-функций" data-toc-modified-id="Определения-функций-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Определения функций</a></span><ul class="toc-item"><li><span><a href="#Загрузка-и-обзор-данных" data-toc-modified-id="Загрузка-и-обзор-данных-1.5.3.1"><span class="toc-item-num">1.5.3.1&nbsp;&nbsp;</span>Загрузка и обзор данных</a></span></li><li><span><a href="#Исправление-данных" data-toc-modified-id="Исправление-данных-1.5.3.2"><span class="toc-item-num">1.5.3.2&nbsp;&nbsp;</span>Исправление данных</a></span></li><li><span><a href="#Отображение-данных" data-toc-modified-id="Отображение-данных-1.5.3.3"><span class="toc-item-num">1.5.3.3&nbsp;&nbsp;</span>Отображение данных</a></span></li><li><span><a href="#A/B-test" data-toc-modified-id="A/B-test-1.5.3.4"><span class="toc-item-num">1.5.3.4&nbsp;&nbsp;</span>A/B-test</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Пропуски-в-данных" data-toc-modified-id="Пропуски-в-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Пропуски в данных</a></span></li><li><span><a href="#Преодоление-неявных-и-явных-дубликатов" data-toc-modified-id="Преодоление-неявных-и-явных-дубликатов-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Преодоление неявных и явных дубликатов</a></span></li><li><span><a href="#Верные-типы-данных" data-toc-modified-id="Верные-типы-данных-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Верные типы данных</a></span></li><li><span><a href="#Добавляем-street" data-toc-modified-id="Добавляем-street-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Добавляем <code>street</code></a></span></li><li><span><a href="#Добавляем-is_continuous" data-toc-modified-id="Добавляем-is_continuous-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Добавляем <code>is_continuous</code></a></span></li><li><span><a href="#Итоги-подготовки-данных" data-toc-modified-id="Итоги-подготовки-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Итоги подготовки данных</a></span></li></ul></li><li><span><a href="#Обобщённые-данные-о-предприятиях-питания" data-toc-modified-id="Обобщённые-данные-о-предприятиях-питания-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обобщённые данные о предприятиях питания</a></span><ul class="toc-item"><li><span><a href="#Распределение-по-расписанию-работы" data-toc-modified-id="Распределение-по-расписанию-работы-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Распределение по расписанию работы</a></span></li><li><span><a href="#Итоги-подготовки-информации-о-расходах:" data-toc-modified-id="Итоги-подготовки-информации-о-расходах:-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Итоги подготовки информации о расходах:</a></span></li></ul></li><li><span><a href="#Портрет-желаемой-кофейни" data-toc-modified-id="Портрет-желаемой-кофейни-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Портрет желаемой кофейни</a></span><ul class="toc-item"><li><span><a href="#Итоги-анализа-расходов" data-toc-modified-id="Итоги-анализа-расходов-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Итоги анализа расходов</a></span></li></ul></li><li><span><a href="#Общие-итоги" data-toc-modified-id="Общие-итоги-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общие итоги</a></span></li></ul></div>

## Описание проекта

### Заказчик

Инвесторы из фонда «Shut Up and Take My Money».

### Задача

Составить описание предприятий общественного питания в Москве. Расположение, тип, меню, цены.

Составить лаконичную презентацию.

Для чего: чтобы выбрать, какое предприятие питания и где открыть.

### Входные данные

Датасет предприятий общепита Москвы из Яндекс.Карт и Яндекс.Бизнеса на лето 2022 года. Информация добавлена пользователями, частично - представителями предприятий, а также собрана автоматически из открытых источников Сети.

#### Термины

При постановке задачи употреблялось слово "заведение". Поскольку речь пойдёт о бизнесе, по-русски - о предпринимательстве, то назовём "заведения" предприятиями. Это слово лучше всего отражает суть: предприятие общепита инвестор воспринимает как что-то, что можно "сделать", предпринять, чтобы заработать. Верные формулировки направят нас к верным выводам.

"Чеком" в общепите часто называют не сам оформленный чековый документ, а стоимость заказа. Мы будем употреблять "стоимость заказа", поскольку смысл тот же, и понятно всем.

#### Описание данных

##### Предприятия питания `places`

- `name` - название;
- `address` - почтовый адрес;
- `category` - тип;
- `hours` - расписание работы;
- `lat` - широта расположения;
- `lng` - долгота расположения;
- `rating` - оценка пользователей Яндекса (высшая - 5.0)
- `price` - категория цен;
- `avg_bill` - диапазон средней стоимости заказа, заказы бывают своего типа у каждого типа предприятия (бокал пива, чашка капучино);
- `middle_avg_bill` - средняя цена заказа в общем (если в `avg_bill` указан именно "Средний счёт")
- `middle_coffee_cup` - средняя цена чашки капучино (если в `avg_bill` указана "Цена одной чашки капучино")
- `chain` - входит ли предприятие в коммерческую сеть (для малых сетей "могут встречаться ошибки" - видимо, предприятия малой сети могут быть отмечены как не входящие в неё, как самостоятельные),
- `district` - городской округ, к которому относится предприятие;
- `seats` - количество посадочных мест.

### План работы

Действия:
- [ ] Импортировать библиотеки, описать функции, которые потребуются:
  - [ ] для расчёта и диаграмм;
  - [ ] для отображения геолокации.
- [ ] Загрузить и изучить таблицы.
- [ ] Подготовить данные:
  - [ ] привести к норме названия столбцов;
  - [ ] обработать дубликаты:
    - [ ] неявные дубликаты значений;
    - [ ] полные дубликаты строк;
  - [ ] привести данные к нужным типам;
  - [ ] обработать пропуски:
    - [ ] выбрать, чем заполнить (или оставить) и почему;
    - [ ] осуществить заполение, если требуется;
    - [ ] описать причины, которые могли привести к пропускам.
  - [ ] Добавить данные:
    - [ ] столбец `street` - названия улиц (из адресов);
    - [ ] столбец `is_24/7` (`bool`) - предприятие непрерывной работы (24 часа в сутки, 7 дней в неделю).
- [ ] Анализировать обобщённые данные:
  - [ ] Какие типы предприятий есть, как распределены? Визуализация этого. Выводы.
  - [ ] Сколько посадочных мест у предприятий каждого типа? Визуализации распределений. Выводы.
  - [ ] Сетевые и самостоятельные предприятия. Каких больше?
  - [ ] В каких категориях сетевые встречаются чаще? Визуализация. Выводы.
  - [ ] Группируем по названиям, чтобы получить 15 самых "популярных" сетей в Москве. (Нужно описать, как оценим "популярность" сети. Это может быть количество предприятий (очень коственный признак популярности), либо количество оценок, либо ещё что-то.) Знакомы ли эти наименования? Что их объединяет? Визуализация количества предприятий каждой сети по округам города.
  - [ ] Какие районы описаны в данных? Общее количество предприятий, и количество предприятий каждой категории по районам. Единая диаграмма - stacked bars.
  - [ ] Средний рейтинг по категориям предприятий. Сильно ли они различаются? Возможна статистическая проверка гипотезы.
  - [ ] Фоновая картограмма среднего рейтинга предприятий каждого района.
  - [ ] Кластерная карта предприятий общепита Москвы.
  - [ ] 15 улиц, на которых больше всего предприятий питания. Диаграмма: какие категории как представлены на этих улицах. Stacked bars.
  - [ ] Улицы, на которых по одной точке питания. Что схожего у этих предприятий?
  - [ ] Медиана средних стоимостей для каждого округа. Картограмма стоимости. Как влияет удалённость от центра города на стоимость?
  - [ ] Какие ещё найдены взаимосвязи? Связь режима работы, расположения и типа предприятия. Связь плохого рейтинга, средней стоимости заказа и категорий.
  - [ ] Описать выводы по обобщённым данным.
- [ ] Идея инвесторов: "крутая" и "доступная" кофейня в Москве (как "Central Perk" в сериале "Друзья"). Конкуренция не считается препятствием. Какие могут быть ные препятствия?
  - [ ] Сколько всего кофеен? В каких районах их больше? Где они размещатся?
  - [ ] Существуют ли круглосуточные кофейни?
  - [ ] Какие у них рейтинги? Как распределены рейтинги кофеен по районам?
  - [ ] Какую стоимость чашки капучино следует установить и почему?
  - [ ] Какие ещё признаки добавить в описание?
  - [ ] В качестве вывода - описать рекомендации.
 [ ] Создать презентацию в PDF, разместить её в облаке, ссылку в ячейке.
  - [ ] что исследовано (указать основные зарактеристики данных);
  - [ ] какие ответы на основные вопросы исследования;
  - [ ] какие гипотезы подтвердились или не подтвердились, отвергнуты;
  - [ ] какие факты обнаружены;
  - [ ] какие предположения выдвинуты;
  - [ ] какие есть рекомендации:
    - [ ] кому,
    - [ ] суть рекомендации.

### Инструменты

Используем [Anaconda](https://www.anaconda.com/), окружение `ds_da_practicum_env.yml` "для Mac OS" (от 2022-03-23).

#### Импорт библиотек

In [1]:
%matplotlib inline

import datetime
import re
import urllib
import json

import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
import folium

#### Определения констант

In [2]:
# set size for all charts
FIGSIZE = (10, 5)

#### Определения функций

##### Загрузка и обзор данных

In [3]:
def load(name, ext='csv', sep=','):
    """Return pandas DataFrame by filename"""
    filename = name + '.' + ext
    try:
        df = pd.read_csv(filename)
    except FileNotFoundError:
        df = pd.read_csv('https://code.s3.yandex.net/datasets/' + filename, sep=sep)
    correct_column_names(df)
    overview(df)
    return df

In [4]:
def overview(data):
    """Display an overview of the data to detect common data problems"""
    display(data.sample(5))
    dupes = data.duplicated().sum()
    dupes_rate = dupes / len(data)
    print(f'Полностью совпадающих строк: {dupes} ({dupes_rate:.2%})\n')
    data.info()
    display(count_nans(data))
    print(get_filled_rows_share(data))

In [5]:
def count_nans(data):
    """Return a pandas DataFrame with the number of missing values"""
    nans = []
    for column in data.columns:
        # count the number of missing values
        missing = data[column].isna().sum()
        # calculate the share of missing values
        part = round(missing * 100 / len(data), 2)
        # count the number of distinct values
        uniques = data[column].sort_values().nunique()
        nans.append([column, missing, part, uniques])

    return pd.DataFrame(
        data=nans, columns=['column', 'na_count', 'na_%', 'unique_count']
    ).sort_values(['na_count', 'unique_count'], ascending=False).set_index('column')

In [6]:
def get_filled_rows_share(df):
    """Return a string containing number of rows with no missing values"""
    share = 1 - df.isna().any(axis=1).sum() / len(df)
    return f'Заполненных строк без пропусков: {share:.02%}'

##### Исправление данных

In [7]:
def correct_column_names(df):
    """Convert column names to valid snake_case in place"""
    df.columns = df.columns.str.strip()
    df.columns = list(map(to_snake_case, df.columns))

def to_snake_case(name):
    """Return name converted to valid snake_case"""    
    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    name = re.sub('__([A-Z])', r'_\1', name)
    name = re.sub('([a-z0-9])([A-Z])', r'\1_\2', name)
    return name.lower()

In [8]:
def fillnaby(df, column, groupby, func='median'):
    """Fill in missing values with group-aggregated values in-place"""
    df[column] = df[column].fillna(
        df.groupby(groupby)[column].transform(func))

##### Отображение данных

In [9]:
def headline(line, width=79):
    """Return the string in plain uppercase text with pseudographics"""
    return f'\n{line.upper():=^{width}}'

##### A/B-test

In [10]:
def center_percentile(data, percentile):
    low = np.percentile(data, (100 - percentile) / 2)
    high = np.percentile(data, (100 + percentile) / 2)
    return low, high

## Подготовка данных

In [11]:
places = load('moscow_places')

name  category                                 address  \
4377    ВьетКафе  ресторан          Москва, улица Земляной Вал, 33   
8196      Смачно  столовая  Москва, Шарикоподшипниковская улица, 1   
3379    Vegolife   бар,паб          Москва, Зоологическая улица, 4   
4865  Coffeewood   кофейня          Москва, Таганская улица, 1/2с2   
6703        Изюм  ресторан             Москва, улица Бутлерова, 12   

                                  district  \
4377    Центральный административный округ   
8196  Юго-Восточный административный округ   
3379    Центральный административный округ   
4865    Центральный административный округ   
6703   Юго-Западный административный округ   

                                                  hours        lat        lng  \
4377                             ежедневно, 10:00–23:00  55.757823  37.659161   
8196                                  пн-пт 08:00–18:00  55.722894  37.665735   
3379               пн-пт 08:30–20:00; сб,вс 10:00–18:00  55.765204  37.580766   
4865  пн-чт 09:00–22:00; пт 09:00–23:00; сб 10:00–23...  55.741385  37.657193   
6703               пн-пт 09:00–18:00; сб,вс 10:00–18:00  55.651582  37.527615   

      rating    price                       avg_bill  middle_avg_bill  \
4377     4.4      NaN       Средний счёт:1000–1500 ₽           1250.0   
8196     4.2   низкие             Средний счёт:150 ₽            150.0   
3379     5.0      NaN                            NaN              NaN   
4865     4.4  средние  Цена чашки капучино:220–270 ₽              NaN   
6703     4.1      NaN                            NaN              NaN   

      middle_coffee_cup  chain  seats  
4377                NaN      1  215.0  
8196                NaN      0   90.0  
3379                NaN      0    NaN  
4865              245.0      0    NaN  
6703                NaN      1   40.0

Полностью совпадающих строк: 0 (0.00%)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8406 entries, 0 to 8405
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   name               8406 non-null   object 
 1   category           8406 non-null   object 
 2   address            8406 non-null   object 
 3   district           8406 non-null   object 
 4   hours              7870 non-null   object 
 5   lat                8406 non-null   float64
 6   lng                8406 non-null   float64
 7   rating             8406 non-null   float64
 8   price              3315 non-null   object 
 9   avg_bill           3816 non-null   object 
 10  middle_avg_bill    3149 non-null   float64
 11  middle_coffee_cup  535 non-null    float64
 12  chain              8406 non-null   int64  
 13  seats              4795 non-null   float64
dtypes: float64(6), int64(1), object(7)
memory usage: 919.5+ KB


na_count   na_%  unique_count
column                                          
middle_coffee_cup      7871  93.64            96
middle_avg_bill        5257  62.54           230
price                  5091  60.56             4
avg_bill               4590  54.60           897
seats                  3611  42.96           229
hours                   536   6.38          1307
lng                       0   0.00          8258
lat                       0   0.00          8209
address                   0   0.00          5753
name                      0   0.00          5614
rating                    0   0.00            41
district                  0   0.00             9
category                  0   0.00             8
chain                     0   0.00             2

Заполненных строк без пропусков: 0.00%


### Пропуски в данных

Полностью заполненных строк - нет. Менее 7% записей содержат цену чашки капучино. Другие цены тоже чаще не указаны. Почти половина точек - без указания количества мест.

При таком количестве пропусков любое их заполнение решительно повлияет на выводы. Лучше убирать отсутствующие результаты при анализе.

Признак "входит в сеть" заполнен, но "может содержать ошибки" - это хуже чем если бы содержал пропуски, ошибки сложнее обнаружить.

Часы работы содержат данные, которые требуется парсить, и затруднительно сказать, какими значениями заполнить остутствующие расписания работы. Прямолинейные статистические методы тут не принесут пользы.

Таким образом, заполнять пропуски представляется не нужным.

Мы можем полагаться на геоданные (координаты, адреса, округа), названия, рейтинг.

### Преодоление неявных и явных дубликатов

Бытовая привычка говорит нам, что если название двух объектов одинаково или слишком похоже - то это одинаковые объекты, например, представители одной сети. Юристы и другие специалисты могут напомнить, что это не всегда так. Вполне могу существовать два предприятия с одинаковыми названиями, никак не связанные между собой. И степень связанности может быть разной - общая администрация, общее юрлицо, общий владелец, общая франшиза, стихийное использование общего громкого названия (особенно если оно не защищено юридически). Поэтому сложно сказать, что именно мы будем считать сетью. Опираться только на названия - не очень надёжно.

Также учитываем, что названия могут быть написаны по-разному просто по стечению обстоятельств. Например, названия предприятий Яндекса записаны в разных источниках по разному, то с точкой, то без неё. Названия записаны по-разному, но предприятия Яндекса определённо не самостоятельны, а объединены.

Основной смысл несут буквы и числа. Именно они произносятся, именно они составляют речь. Пробелы, знаки препинания, служебные символы, регистр - важны гораздо меньше, и именно они обычно различаются при написании названия одного и того же объекта, поскольку в речи они трудно различимы.

Поэтому уникальность строки будем определять не по всему её содержимому, а только оставив в строке буквы и цифры, без всего, без пробелов. Это тоже вполне читаемый текст, это и есть суть текста, суть названия.

Если строка содержит те же буквы и цифры в том же порядке, как другая строка - они считаются одинаковыми. Сперва поищем, сколько одинаковых по смыслу строк в текстовых столбцах.

In [12]:
def get_meaning_pattern(text):
    # имеющими значение будем считать только буквы и цифры
    return ''.join(char for char in text.lower() if char.isalpha() or char.isdigit())

# проверяем все текстовые поля,
# есть ли в них строки, различающиеся не смыслом, а формой записи
for column in places.select_dtypes(include='object').columns:
    values = places[column].dropna().unique()
    patterns = set(map(get_meaning_pattern, values))
    if len(patterns) != len(values):
        print('Столбец {} содержит неодинаковое количество различных значений:\n'
              'по смыслу их {} разных, по форме записи {} разных'.format(
                  column, len(patterns), len(values)))
        print(values + '\n')

Столбец name содержит неодинаковое количество различных значений:
по смыслу их 5454 разных, по форме записи 5614 разных
['WoWфли\n' 'Четыре комнаты\n' 'Хазри\n' ... 'Миславнес\n' 'Самовар\n'
 'Kebab Time\n']
Столбец address содержит неодинаковое количество различных значений:
по смыслу их 5750 разных, по форме записи 5753 разных
['Москва, улица Дыбенко, 7/1\n' 'Москва, улица Дыбенко, 36, корп. 1\n'
 'Москва, Клязьминская улица, 15\n' ...
 'Москва, улица Лобачевского, 52, корп. 1\n'
 'Москва, Болотниковская улица, 52, корп. 2\n'
 'Москва, Чонгарский бульвар, 26А, корп. 1\n']


Подготовим функцию, которая, получив строку таблицы, возвращает текст из выбранной колонки. Если ранее уже встречалась другая формулировка этого же текста - то вернёт ту, ранее встречавшуюся. Смыслы и соответствующие им формулировки накапливаются во внешнем словаре. Поэтому перед использованием функции словарь надо создавать (пустым).

In [13]:
def get_mean_text(row, column):
    pattern = get_meaning_pattern(row[column])    
    if pattern in patterns.keys() and row[column] != patterns[pattern]:
        print(f"{row[column]} -> {patterns[pattern]}")
        return patterns[pattern]
    else:
        patterns[pattern] = row[column]
        return row[column]

Проверка на уникальность показала, что 'names' и 'addresses' содержат неявные дубликаты. Поскольку дальнейшие исследования будут использовать названия - то исходные столбцы менять не будем, а добавим дополнительные 2 столбца, с "подразумеваемыми" названиями и адресами.

In [14]:
patterns = dict()
column = 'name'
places[f'mean_{column}'] = places.apply(
    lambda row: get_mean_text(row, column), axis=1)

Хинкали-Gали! -> Хинкали Gали!
Хинкали-Gали! -> Хинкали Gали!
Доминос пицца -> Домино'с Пицца
Хинкали-Gали! -> Хинкали Gали!
Meat Лав -> MeatЛав
Донер кебаб -> Донер Кебаб
Чай хана халал -> Чайхана Халал
В парке вкуснее! -> В парке вкуснее
Кафе-бар -> Кафе бар
Чайхона №1 -> Чайхона № 1
Кафе-Столовая -> Кафе-столовая
Кафе-бар -> Кафе бар
МСК lounge -> МСК Lounge
Хинкали Gали -> Хинкали Gали!
Франклинс бургер -> Франклинс Бургер
Wild Bean Cafe -> Wild bean cafe
Wild Bean Cafe -> Wild bean cafe
Яндекс.Лавка -> Яндекс Лавка
Wild Bean Cafe -> Wild bean cafe
Чайхана халяль -> Чайхана Халяль
Новая столовая -> Новая Столовая
More Poke -> More poke
Кафе-бар -> Кафе бар
SeDelice -> Sedelice
Shake up -> ShakeUp
Pho Bo -> PhoBo
Хинкали - Gали! -> Хинкали Gали!
Ача-Чача -> Ача-чача
Важная Персона -> Важная персона
Jeffrey's Coffeeshop -> Jeffrey’s Coffeeshop
It's СоТ-Кофейня -> It's СоТ - Кофейня
Лепим и Варим -> Лепим и варим
Чайхана халяль -> Чайхана Халяль
Кафе-бар -> Кафе бар
Wild Bean Cafe -> 

In [15]:
patterns = dict()
column = 'address'
places[f'mean_{column}'] = places.apply(
    lambda row: get_mean_text(row, column), axis=1)

Москва, парк Ангарские Пруды -> Москва, парк Ангарские пруды
Москва, Кутузовский проспект, 17 -> Москва, Кутузовский проспект, 1/7
Москва, Кутузовский проспект, 17 -> Москва, Кутузовский проспект, 1/7
Москва, Кутузовский проспект, 17 -> Москва, Кутузовский проспект, 1/7
Москва, улица Академика Варги, 8 к 1 -> Москва, улица Академика Варги, 8к1
Москва, улица Академика Варги, 8 к 1 -> Москва, улица Академика Варги, 8к1


In [16]:
print('Различных названий:', len(places[f'mean_{column}'].unique()))

Различных названий: 5750


In [17]:
print('Различных адресов:', len(places[f'mean_{column}'].unique()))

Различных адресов: 5750


Полученные адреса и названия различны не только по форме, но и по смыслу. Поэтому посчитаем дубликаты ещё раз, считая уникальным сочетание подразумеваемого названия и подразумеваемого адреса.

In [18]:
places[places[['mean_name', 'lat', 'lng']].duplicated()][['mean_name', 'lat', 'lng']]

mean_name        lat        lng
1511  More poke  55.806307  37.497566

In [19]:
dupes = places[
    places[['mean_name', 'mean_address']].duplicated()
][['mean_name', 'mean_address']]

display(dupes)

mean_name  \
215                        Кафе   
600             В парке вкуснее   
1511                  More poke   
2420  Раковарня Клешни и Хвосты   
3109            Хлеб да Выпечка   
4780                    Cafe 13   

                                           mean_address  
215                        Москва, парк Ангарские пруды  
600   Москва, Северный административный округ, район...  
1511            Москва, Волоколамское шоссе, 11, стр. 2  
2420                         Москва, проспект Мира, 118  
3109                        Москва, Ярцевская улица, 19  
4780               Москва, Мясницкая улица, 13, стр. 11

Видим, что вместе с записью 1511 - у нас 6 дубликатов. Убедимся, что это одни и те же рестораны, упомянутые дважды. Кроме того, взглянем, какую информацию можно объединить из двух строк в одну.

In [20]:
def select_rows(keys):
    display(places[
        (places['mean_name'] == keys['mean_name'])
        & (places['mean_address'] == keys['mean_address'])
    ])

dupes.apply(select_rows, axis=1)

name category                       address  \
189  Кафе     кафе  Москва, парк Ангарские Пруды   
215  Кафе     кафе  Москва, парк Ангарские пруды   

                            district                   hours        lat  \
189  Северный административный округ  ежедневно, 09:00–23:00  55.880327   
215  Северный административный округ  ежедневно, 10:00–22:00  55.881438   

           lng  rating price avg_bill  middle_avg_bill  middle_coffee_cup  \
189  37.530786     3.2   NaN      NaN              NaN                NaN   
215  37.531848     3.2   NaN      NaN              NaN                NaN   

     chain  seats mean_name                  mean_address  
189      0    NaN      Кафе  Москва, парк Ангарские пруды  
215      0    NaN      Кафе  Москва, парк Ангарские пруды

name category  \
599  В парке вкуснее!     кафе   
600   В парке вкуснее  кофейня   

                                               address  \
599  Москва, Северный административный округ, район...   
600  Москва, Северный административный округ, район...   

                            district                   hours        lat  \
599  Северный административный округ  ежедневно, 10:00–21:00  55.854571   
600  Северный административный округ  ежедневно, 10:00–21:00  55.851985   

           lng  rating price avg_bill  middle_avg_bill  middle_coffee_cup  \
599  37.487254     2.2   NaN      NaN              NaN                NaN   
600  37.478492     2.2   NaN      NaN              NaN                NaN   

     chain  seats        mean_name  \
599      0    NaN  В парке вкуснее   
600      1    NaN  В парке вкуснее   

                                          mean_address  
599  Москва, Северный административный округ, район...  
600  Москва, Северный административный округ, район...

name  category                                  address  \
1430  More poke  ресторан  Москва, Волоколамское шоссе, 11, стр. 2   
1511  More Poke  ресторан  Москва, Волоколамское шоссе, 11, стр. 2   

                             district  \
1430  Северный административный округ   
1511  Северный административный округ   

                                                  hours        lat        lng  \
1430                             ежедневно, 09:00–21:00  55.806307  37.497566   
1511  пн-чт 09:00–18:00; пт,сб 09:00–21:00; вс 09:00...  55.806307  37.497566   

      rating price avg_bill  middle_avg_bill  middle_coffee_cup  chain  seats  \
1430     4.2   NaN      NaN              NaN                NaN      0  188.0   
1511     4.2   NaN      NaN              NaN                NaN      1  188.0   

      mean_name                             mean_address  
1430  More poke  Москва, Волоколамское шоссе, 11, стр. 2  
1511  More poke  Москва, Волоколамское шоссе, 11, стр. 2

name  category                     address  \
2211  Раковарня Клешни и Хвосты  ресторан  Москва, проспект Мира, 118   
2420  Раковарня Клешни и хвосты   бар,паб  Москва, проспект Мира, 118   

                                     district  \
2211  Северо-Восточный административный округ   
2420  Северо-Восточный административный округ   

                                                  hours        lat        lng  \
2211                             ежедневно, 12:00–00:00  55.810553  37.638161   
2420  пн-чт 12:00–00:00; пт,сб 12:00–01:00; вс 12:00...  55.810677  37.638379   

      rating price avg_bill  middle_avg_bill  middle_coffee_cup  chain  seats  \
2211     4.4   NaN      NaN              NaN                NaN      0  150.0   
2420     4.4   NaN      NaN              NaN                NaN      1  150.0   

                      mean_name                mean_address  
2211  Раковарня Клешни и Хвосты  Москва, проспект Мира, 118  
2420  Раковарня Клешни и Хвосты  Москва, проспект Мира, 118

name  category                      address  \
3091  Хлеб да Выпечка  булочная  Москва, Ярцевская улица, 19   
3109  Хлеб да выпечка      кафе  Москва, Ярцевская улица, 19   

                             district                   hours        lat  \
3091  Западный административный округ  ежедневно, 09:00–22:00  55.738886   
3109  Западный административный округ                     NaN  55.738449   

            lng  rating price avg_bill  middle_avg_bill  middle_coffee_cup  \
3091  37.411648     4.1   NaN      NaN              NaN                NaN   
3109  37.410937     4.1   NaN      NaN              NaN                NaN   

      chain  seats        mean_name                 mean_address  
3091      1  276.0  Хлеб да Выпечка  Москва, Ярцевская улица, 19  
3109      0  276.0  Хлеб да Выпечка  Москва, Ярцевская улица, 19

name  category                               address  \
4613  Cafe 13      кафе  Москва, Мясницкая улица, 13, стр. 11   
4780   Cafe13  ресторан  Москва, Мясницкая улица, 13, стр. 11   

                                district                              hours  \
4613  Центральный административный округ  пн-чт 10:00–22:00; пт 10:00–19:00   
4780  Центральный административный округ  пн-чт 10:00–22:00; пт 10:00–18:00   

            lat        lng  rating price avg_bill  middle_avg_bill  \
4613  55.762779  37.633079     4.3   NaN      NaN              NaN   
4780  55.762748  37.633176     4.3   NaN      NaN              NaN   

      middle_coffee_cup  chain  seats mean_name  \
4613                NaN      0  200.0   Cafe 13   
4780                NaN      0  200.0   Cafe 13   

                              mean_address  
4613  Москва, Мясницкая улица, 13, стр. 11  
4780  Москва, Мясницкая улица, 13, стр. 11

215     None
600     None
1511    None
2420    None
3109    None
4780    None
dtype: object

В каждой паре записей совпадают рейтинги и количество мест. Дистанция между ними по координатам - от нескольких метров до нескольких сотен метров. Сравнив данные с информацией, предоставляемой сейчас пользователям Яндекс.Картами, не удаётся определить, кака информация актуальнее - в каждом случае на картах присутствуют обе записи. Получать срежднюю величину из пары, поэтому удаляем первые результаты - скорее всего, они старше и менее актуальны.

In [21]:
places = places.drop([189, 599, 1430, 2211, 3091, 4613]).reset_index(drop=True)

### Верные типы данных

Количество мест должно быть целым числом.

In [22]:
places['seats'] = places['seats'].astype({'seats': 'Int64'})
places.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8400 entries, 0 to 8399
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   name               8400 non-null   object 
 1   category           8400 non-null   object 
 2   address            8400 non-null   object 
 3   district           8400 non-null   object 
 4   hours              7864 non-null   object 
 5   lat                8400 non-null   float64
 6   lng                8400 non-null   float64
 7   rating             8400 non-null   float64
 8   price              3315 non-null   object 
 9   avg_bill           3816 non-null   object 
 10  middle_avg_bill    3149 non-null   float64
 11  middle_coffee_cup  535 non-null    float64
 12  chain              8400 non-null   int64  
 13  seats              4791 non-null   Int64  
 14  mean_name          8400 non-null   object 
 15  mean_address       8400 non-null   object 
dtypes: Int64(1), float64(5),

### Добавляем `street`

... чтобы проанализировать, на каких улицах есть общепит.

In [23]:
street_attributes = [
    'улица', 'ул.', 'просп', 'проезд', 'бул.', 'бульвар', 'б-р',
    'шоссе', 'площ',
]

def get_street(place):
    addr = place['address'].split(', ')

    # иногда в адресах вторым идёт какой-нибудь район или посёлок,
    # а в конце идёт не только дом, но и строение,
    # поэтому будем искать улицу не по позиции в адресе,
    # а очевидным способом - по упоминанию слова "улица" или аналогичного
    for part in addr:
        for attr in street_attributes:
            if attr in part:
                return part

    # если не найдём - вернём второй элемент адреса
    return addr[1]

places['street'] = places.apply(get_street, axis=1)
places.sample(10)[['address', 'street']]

address  \
4026                          Москва, Брянская улица, 5   
2406           Москва, Верхняя Красносельская улица, 10   
725                 Москва, Бескудниковский бульвар, 13   
1744  Москва, Северный административный округ, улица...   
5226              Москва, Летниковская улица, 2, стр. 2   
2009                 Москва, Часовая улица, 16, стр. 21   
6171                     Москва, Духовской переулок, 17   
1462                  Москва, 3-я Хорошёвская улица, 12   
7414            Москва, МКАД, 14-й километр, 2, стр. 64   
6844                   Москва, улица Островитянова, 32А   

                            street  
4026                Брянская улица  
2406  Верхняя Красносельская улица  
725        Бескудниковский бульвар  
1744                 улица 8 Марта  
5226            Летниковская улица  
2009                 Часовая улица  
6171            Духовской переулок  
1462         3-я Хорошёвская улица  
7414                          МКАД  
6844           улица Островитянова

### Добавляем `is_continuous`

... чтобы проверить, выделяются ли рестораны, работающие без перерывов, чем-нибудь особенным.

In [24]:
# NaN по умолчанию превращается в False
places['is_continuous'] = places['hours'] == 'ежедневно, круглосуточно'
# А нужно, чтобы оставался третий вариант - "неизвестно"
places.loc[places['hours'].isna(), 'is_continuous'] = None

places[['hours', 'is_continuous']].sample(10)

hours is_continuous
7552                                                NaN           NaN
8325                             ежедневно, 11:00–02:00         False
5529                             ежедневно, 09:00–20:00         False
1892                                  пн-пт 09:00–18:00         False
4759                             ежедневно, 12:00–23:00         False
7703                             ежедневно, 07:00–23:00         False
739                пн-пт 08:00–22:00; сб,вс 09:00–22:00         False
6923                             ежедневно, 10:00–02:00         False
5843  пн-чт 10:00–22:00; пт,сб 10:00–23:00; вс 10:00...         False
3297                             ежедневно, 10:00–22:00         False

### Итоги подготовки данных

- удалено 6 дубликатов;
- зафиксированы пропуски в столбцах: middle_coffee_cup (93.64%), middle_avg_bill (62.54%), price (60.56%), avg_bill (54.60%), seats	(42,96%), и относительно мало пропусков в hours (6,38%);
- добавлены столбцы `street` и `is_continuous`.

In [27]:
places.sample(5)

name  category                                          address  \
4326     Оджахури  ресторан  Москва, Садовая-Черногрязская улица, 22, стр. 1   
8087      Вахтанг  ресторан              Москва, улица Ленинская Слобода, 26   
6637  Кофе&сласти   кофейня             Москва, улица Покрышкина, 8, корп. 1   
511   Шоколадница      кафе         Москва, Ленинградское шоссе, 16А, стр. 4   
1361  Шоколадница   кофейня                      Москва, Щукинская улица, 42   

                                    district  \
4326      Центральный административный округ   
8087            Южный административный округ   
6637         Западный административный округ   
511          Северный административный округ   
1361  Северо-Западный административный округ   

                                                  hours        lat        lng  \
4326  пн-чт 12:00–00:00; пт,сб 12:00–02:00; вс 12:00...  55.764406  37.655277   
8087  пн-чт 10:00–22:00; пт 10:00–23:00; сб 11:00–23...  55.709017  37.652758   
6637                             ежедневно, 07:00–21:00  55.665961  37.469625   
511                              ежедневно, 10:00–23:00  55.823176  37.497703   
1361                             ежедневно, 10:00–22:00  55.809257  37.464076   

      rating    price                       avg_bill  middle_avg_bill  \
4326     4.6  высокие         Средний счёт:от 3500 ₽           3500.0   
8087     3.4      NaN                            NaN              NaN   
6637     4.1      NaN  Цена чашки капучино:200–300 ₽              NaN   
511      4.1      NaN                            NaN              NaN   
1361     4.3  средние        Средний счёт:700–1000 ₽            850.0   

      middle_coffee_cup  chain  seats    mean_name  \
4326                NaN      0     98     Оджахури   
8087                NaN      0    150      Вахтанг   
6637              250.0      0    100  Кофе&сласти   
511                 NaN      1    230  Шоколадница   
1361                NaN      1    150  Шоколадница   

                                         mean_address  \
4326  Москва, Садовая-Черногрязская улица, 22, стр. 1   
8087              Москва, улица Ленинская Слобода, 26   
6637             Москва, улица Покрышкина, 8, корп. 1   
511          Москва, Ленинградское шоссе, 16А, стр. 4   
1361                      Москва, Щукинская улица, 42   

                           street is_continuous  
4326  Садовая-Черногрязская улица         False  
8087      улица Ленинская Слобода         False  
6637             улица Покрышкина         False  
511           Ленинградское шоссе         False  
1361              Щукинская улица         False

## Обобщённые данные о предприятиях питания

### Распределение по расписанию работы

In [25]:
common_hours = (
    places.groupby('hours').agg({'name': 'count'})
    .rename(columns={'name': 'count'})
    .sort_values('count', ascending=False)
)

common_hours.head(10)

count
hours                          
ежедневно, 10:00–22:00      759
ежедневно, круглосуточно    730
ежедневно, 11:00–23:00      396
ежедневно, 10:00–23:00      310
ежедневно, 12:00–00:00      253
ежедневно, 09:00–21:00      203
ежедневно, 09:00–22:00      183
ежедневно, 12:00–23:00      178
ежедневно, 08:00–23:00      160
ежедневно, 08:00–22:00      148

### Итоги подготовки информации о расходах:

- Первый;
- Второй;
- Третий.

In [26]:
district_boundaries = json.loads(urllib.request.urlopen(
    'https://code.s3.yandex.net/data-analyst/admin_level_geomap.geojson'
).read())


## Портрет желаемой кофейни

### Итоги анализа расходов

- Первый;
- Второй;
- Третий.

## Общие итоги

- Исследовано:
- **Подтверждённые выводы:**
- Также обнаружено, что:
- Предположения: